File: Dynamic Programming/Integer Break/solution_top_down.py
class Solution:
    def integerBreak(self, e: int) -> int:
        dp = [None for i in range(e + 1)]
        def dfs(n):
            if n == 1: return 1
            if dp[n] != None: return dp[n]
            res = n
            if n == e: res = 0
            for i in range(1, n):
                res = max(res, i * dfs(n - i))
            dp[n] = res
            return res
        return dfs(e)
        

File: Dynamic Programming/Integer Break/solution_top_down.ts
function integerBreak(n: number): number {
    let given = n
    let dp = Array(n)
    let dfs = (n) => {
        if(n === 1) return 1
        if(dp[n] !== undefined) return dp[n]
        let result = n
        if(n === given) result = 0
        for(let i = 1; i < n; i++) {
            result = Math.max(result, dfs(i) * dfs(n - i))
        }
        dp[n] = result
        return result
    }
    return dfs(n)
    
};